tests/SimplePacker.hpp

```diff
@@ -0,0 +1,293 @@
+    //! \param data The array to pack
+    //! \param n Length of array
+    //! \param buffer Buffer to pack into
+    //! \param position Position in buffer to use
+    static void pack(const T* data,
+                     std::size_t n,
+                     std::vector<char>& buffer,
+                     int& position)
+    {
+        std::memcpy(&buffer[position], data, n*sizeof(T));
+        position += n*sizeof(T);
+    }
+
+    //! \brief Unpack a POD.
+    //! \param data The variable to unpack
+    //! \param buffer Buffer to unpack from
+    //! \param position Position in buffer to use
+    static void unpack(T& data,
+                       std::vector<char>& buffer,
+                       int& position)
+    {
+        unpack(&data, 1, buffer, position);
+    }
+
+    //! \brief Unpack an array of POD.
+    //! \param data The array to unpack
+    //! \param n Length of array
+    //! \param buffer Buffer to unpack from
+    //! \param position Position in buffer to use
+    static void unpack(T* data,
+                       std::size_t n,
+                       std::vector<char>& buffer,
+                       int& position)
+    {
+        std::memcpy(data, &buffer[position], n*sizeof(T));
+        position += n*sizeof(T);
+    }
+};
+
+//! \brief Default handling for unsupported types.
+template<class T>
+struct Packing<false,T>
+{
+    static std::size_t packSize(const T&)
+    {
+        static_assert(!std::is_same_v<T,T>, "Packing not supported for type");
+        return 0;
+    }
+
+    static void pack(const T&, std::vector<char>&, int&)
+    {
+        static_assert(!std::is_same_v<T,T>, "Packing not supported for type");
+    }
+
+    static void unpack(T&, std::vector<char>&, int&)
+    {
+        static_assert(!std::is_same_v<T,T>, "Packing not supported for type");
+    }
+};
+
+//! \brief Specialization for std::bitset
+template <std::size_t Size>
+struct Packing<false,std::bitset<Size>>
+{
+    static std::size_t packSize(const std::bitset<Size>& data)
+    {
+        return Packing<true,unsigned long long>::packSize(data.to_ullong());
+    }
+
+    static void pack(const std::bitset<Size>& data,
+                     std::vector<char>& buffer, int& position)
+    {
+        Packing<true,unsigned long long>::pack(data.to_ullong(), buffer, position);
+    }
+
+    static void unpack(std::bitset<Size>& data,
+                       std::vector<char>& buffer, int& position)
+    {
+        unsigned long long d;
+        Packing<true,unsigned long long>::unpack(d, buffer, position);
+        data = std::bitset<Size>(d);
+    }
+};
+
+template<>
+struct Packing<false,std::string>
+{
+    static std::size_t packSize(const std::string& data)
+    {
+        return sizeof(std::size_t) + strlen(data.c_str());
+    }
+
+    static void pack(const std::string& data,
+                     std::vector<char>& buffer, int& position)
+    {
+        std::size_t length = strlen(data.c_str());
```

Comment:
Better hope we never get strings with embedded nuls.  I guess it's okay for testing purposes, but is there a reason not to use `std::string::size()` here?

Reply:
well yes, string::size doesn't reflect underlying data size for multibyte encodings, ref https://github.com/OPM/opm-simulators/pull/4107#discussion_r971700667

Reply:
indeed. i should have checked the documentation again, i also remembered it the wrong way around. .size() will do the correct thing.

Reply:
:flushed: Yeah, maybe I was confused. I am always reading the basic_string documentation and get nervous about CharT. Indeed as std::string uses char there should be no concerns. Sorry for the noise.